File: 01-Intro-Langchain/intro-langchain-python/examples/4-gerenciamento-de-memoria/1-armazenamento-de-historico.py
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.chat_history import InMemoryChatMessageHistory
from langchain_core.runnables import RunnableConfig, RunnableWithMessageHistory
from dotenv import load_dotenv
from pydantic import SecretStr
import os

load_dotenv()

prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a helpful assistant."),
    MessagesPlaceholder(variable_name="history"),
    ("human", "{input}"),
])

chat_model = ChatOpenAI(
    model="gpt-4o", 
    temperature=0.9,
    base_url=os.environ["GITHUB_MODELS_ENDPOINT"],
    api_key=SecretStr(os.environ["GITHUB_MODELS_TOKEN"])
)

chain = prompt | chat_model

session_store: dict[str, InMemoryChatMessageHistory] = {}

def get_session_history(session_id: str) -> InMemoryChatMessageHistory:
  if session_id not in session_store:
    session_store[session_id] = InMemoryChatMessageHistory()
  return session_store[session_id]

conversational_chain = RunnableWithMessageHistory(
    chain,
    get_session_history,
    input_messages_key="input",
    history_messages_key="history",
)

config: RunnableConfig = {"configurable": {"session_id": "demo-session"}}

response1 = conversational_chain.invoke({"input": "Hello, my name is Glaucia. How are you?"}, config=config)
print("Assistant: ", response1.content)
print("-"*30)

response2 = conversational_chain.invoke({"input": "Can you repeat my name?"}, config=config)
print("Assistant: ", response2.content)
print("-"*30)

response3 = conversational_chain.invoke({"input": "Can you repeat my name in a motivation phrase?"}, config=config)
print("Assistant: ", response3.content)
print("-"*30)

File: 01-Intro-Langchain/intro-langchain-typescript/src/examples/4-gerenciamento-de-memoria/1-armazenamento-de-historico.ts
import { ChatOpenAI } from "@langchain/openai";
import {
  ChatPromptTemplate,
  MessagesPlaceholder,
} from "@langchain/core/prompts";
import { InMemoryChatMessageHistory } from "@langchain/core/chat_history";
import {
  RunnableConfig,
  RunnableWithMessageHistory,
} from "@langchain/core/runnables";
import { loadEnvironment, getApiKey, getEndpoint } from "../../utils/helpers";

loadEnvironment();

const prompt = ChatPromptTemplate.fromMessages([
  ["system", "You are a helpful assistant."],
  new MessagesPlaceholder("history"),
  ["human", "{input}"],
]);

const chatModel = new ChatOpenAI({
  model: "gpt-4o",
  temperature: 0.9,
  configuration: {
    baseURL: getEndpoint("GITHUB_MODELS_ENDPOINT"),
    apiKey: getApiKey("GITHUB_MODELS_TOKEN"),
  },
});

const chain = prompt.pipe(chatModel);

const sessionStore = new Map<string, InMemoryChatMessageHistory>();

function getSessionHistory(sessionId: string): InMemoryChatMessageHistory {
  if (!sessionStore.has(sessionId)) {
    sessionStore.set(sessionId, new InMemoryChatMessageHistory());
  }

  return sessionStore.get(sessionId)!;
}

const conversationalChain = new RunnableWithMessageHistory({
  runnable: chain,
  getMessageHistory: getSessionHistory,
  inputMessagesKey: "input",
  historyMessagesKey: "history",
});

const config: RunnableConfig = {
  configurable: { sessionId: "demo-session" },
};

async function main(): Promise<void> {
  const response1 = await conversationalChain.invoke(
    { input: "Hello, my name is Glaucia. How are you?" },
    config
  );
  console.log("Assistant:", response1.content);
  console.log("-".repeat(30));

  const response2 = await conversationalChain.invoke(
    { input: "Can you repeat my name?" },
    config
  );
  console.log("Assistant:", response2.content);
  console.log("-".repeat(30));

  const response3 = await conversationalChain.invoke(
    { input: "Can you repeat my name in a motivation phrase?" },
    config
  );
  console.log("Assistant:", response3.content);
  console.log("-".repeat(30));
}

void main();
